### DISCUSSION TITLE
clang-format and clang-tidy versions

### DISCUSSION THREAD

**Discussion by ghost** [ihhub/fheroes2#4445] [Q&A] [answered]
Hello everyone,

First of all - the project is really great! I don't know which I enjoy more: playing the game or reading the code :)

What is the proper way of auto-formatting the code with `clang-format`? Which version of `clang-format` should I use?
When doing `clang-format` with provided configuration I get many changes in original code. 

E.g. (file `fheroes2/battle/battle_interface.cpp`) from
`std::vector<std::pair<LightningPoint, LightningPoint> > lines;`
to
`std::vector<std::pair<LightningPoint, LightningPoint>> lines;`

Or from
`static const uint8_t contourColorTable[] = {108, 115, 122, 129, 122, 115};`
to
`static const uint8_t contourColorTable[] = { 108, 115, 122, 129, 122, 115 };`

Similarly `clang-tidy` displays a lot of warnings like
`warning: redundant void argument list in function definition [modernize-redundant-void-arg]`
or
`warning: parameter 'offset' is unused [misc-unused-parameters]`

I work on Arch Linux. The versions of tools I use:
```
$ clang-format --version
clang-format version 12.0.1

$ clang-tidy --version
LLVM (http://llvm.org/):
  LLVM version 12.0.1
  Optimized build.
  Default target: x86_64-pc-linux-gnu
  Host CPU: skylake
```

Kind regards.
Mateusz

**Comment by ihhub** [comment, marked as answer]
Hi @seacorn , clang-format was upgraded from C++03 to C++11 sometime a year ago. This means that the code which was formatted by old rules are not suitable for new rules. That's why clang-format complains about the old code. Your version of the tool is absolutely correct to use.

The same applies for clang-tidy: we still have a lot of legacy code containing many problematic places for code quality. My advice: if you don't modify a particular line of the code where clang-tidy complains then ignore it. In our CI/CD pipeline we use clang-tidy only on modified code.

Don't worry about these issues with the tools. You can open a pull request and made subsequent commits if needed. All commits are going to be squashed at the time of pull request merging.

I hope I answered all your questions.

**Reply by a1exsh** [reply]
I'm just using this script as `.git/hook/pre-commit`:
```.sh
#!/bin/bash
exec bash ./script/tools/check_code_format.sh
```

It catches formatting problems in the changed/added lines quite reliably.

**Reply by ghost** [reply]
@ihhub , yes, everything is clear now - thanks. 

@a1exsh , thanks for the hint. I had to create a link `clang-format-diff -> /usr/share/clang/clang-format-diff.py` because in my case it was not available in `$PATH`. Perhaps a short comment in the script about the origin of `clang-format-diff` would be helpful (I didn't know about this tool).
